### PR TITLE
Yolk Heroes Initial Commit

### PR DESCRIPTION
**Pull request by EzDzzIt**
# New Port for { Game Title }

## Game Information
- **Title**: Yolk Heroes: A Long Tamago
- **URL**: n/a

### CFW Tests
Ensure your game has been tested on all major CFWs:
- [x ] AmberELEC
- [x ] ArkOS
- [x ] ROCKNIX
- [x ] muOS

### Resolution Tests
Test all major resolutions:
- [ ] 480x320
- [x ] 640x480
- [ ] 720x720 (RGB30)
- [x ] Higher resolutions (e.g., 1280x720)

## File Structure
- Your port should have the following structure:
  - portname/
    - port.json
    - README.md
    - screenshot.jpg
    - cover.jpg
    - gameinfo.xml
    - Port Name.sh
    - portname/
      - <portfiles here>

## Additional Resources
For an in-depth guide on creating a pull request, refer to: [PortMaster Game Packaging Guide](https://portmaster.games/packaging.html#creating-a-pull-request)

### CODE REVIEW

File: ports/yolkheroes/README.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+## Notes
+
+Please support the developers at https://store.steampowered.com/app/2417850/?snr=1_5_9__205 or https://pixelwestern.itch.io/yolk-hero. Purchasing the game is mandatory to run it.
+
+*Someone has to protect this realm... is it you? Become a guardian spirit and raise your very own Hero to stand against the Dark Lord.*
+
+Yolk Heroes is a Tamagotchi-style idle game where the player raises an Elf to become a hero. Think Digimon Tamagotchi meets Princess Maker as an idle game.
```
Comment (Cebion) [inline ports/yolkheroes/README.md:7]: Description is done in the port.json for the website and the PM app.

File: ports/yolkheroes/README.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+The Gameboy-style art and screen scaling make it a good candidate for a portable platform.
+
+## Controls
+
+| Button | Action |
```
Comment (JeodC) [inline ports/yolkheroes/README.md:13]: Buttons should tell the user what they do, not tell what keyboard key it maps to.

File: ports/yolkheroes/Yolk Heroes.sh [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+export XDG_CONFIG_HOME="$CONFDIR"
+export XDG_DATA_HOME="$CONFDIR"
+
+# Keyboard Entry; edit name here
+export TEXTINPUTPRESET="NAME"         # defines preset text to insert
+export TEXTINPUTINTERACTIVE="Y"        # enables interactive text input mode
+export TEXTINPUTNOAUTOCAPITALS="Y"     # disables automatic capitalisation of first letter of words in interactive text input mode
+export TEXTINPUTADDEXTRASYMBOLS="Y"    # enables additional symbols for interactive text input
+
+# # .check for patched .pck ...
```
Comment (Cebion) [inline ports/yolkheroes/Yolk Heroes.sh:37]: Not used? Can be deleted?

Comment (JeodC) [inline ports/yolkheroes/Yolk Heroes.sh:37]: What is this for? If we aren't using it, it shouldn't be in the final launchscript.

File: ports/yolkheroes/port.json [hunk 4]
```diff
@@ -0,0 +1,32 @@
+{
+  "version": 3,
+  "name": "YolkHeroes.zip",
```
Comment (Cebion) [inline ports/yolkheroes/port.json:3]: zip name needs to be match the foldername. Both lowercase

File: ports/yolkheroes/cover.png
Comment (JeodC) [inline ports/yolkheroes/cover.png:1]: I can design a cover if you tell me what screenshot you want to use. If you go with this it has to be 640x480, so letterboxing.

File: ports/yolkheroes/gameinfo.xml [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+<gameList>
+  <game>
+    <path>./YolkHeroes.sh</path>
+    <name>Yolk Heroes: A Long Tomago</name>
+    <desc>Tamagotchi-style Fantasy Idle game.</desc>
+    <releasedate>20240520T000000</releasedate>
+    <developer>14 Hours Productions</developer>
+    <publisher>14 Hours Productions, Astrolabe Games</publisher>
+    <genre>Idle</genre>
+    <image>./yolkheroes/screenshot.png</image>
```
Comment (JeodC) [inline ports/yolkheroes/gameinfo.xml:11]: cover.png

File: ports/yolkheroes/port.json [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  "items": [
+    "YolkHeroes.sh",
```
Comment (JeodC) [inline ports/yolkheroes/port.json:5]: No spaces?

File: ports/yolkheroes/yolkheroes/yolkheroes.gptk [hunk 7]
```diff
@@ -0,0 +1,16 @@
+up = up
+down = down
+left = left
+right = right
+
+a = z
+b = x
+x = tab
```
Comment (Cebion) [inline ports/yolkheroes/yolkheroes/yolkheroes.gptk:8]: In the readme both A and x is Z.
which one is correct?

**Review by Cebion — DISMISSED** [review]

File: ports/yolkheroes/gameinfo.xml [hunk 8]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<gameList>
+  <game>
+    <path>./YolkHeroes.sh</path>
+    <name>Yolk Heroes: A Long Tomago</name>
```
Comment (JeodC) [inline ports/yolkheroes/gameinfo.xml:5]: Is it Tomago or Tamago?

File: ports/yolkheroes/port.json [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    "yolkheroes/"
+  ],
+  "items_opt": [],
+  "attr": {
+    "title": "Yolk Heroes: A Long Tomago",
```
Comment (JeodC) [inline ports/yolkheroes/port.json:10]: Tomago or Tamago?

File: ports/yolkheroes/port.json [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    "porter": [
+      "nate_____"
+    ],
+    "desc": "A cute Tamagotchi idle game from 14 Hours Productions.",
+    "desc_md": null,
+    "inst": "Purchase game from https://store.steampowered.com/app/2417850/?snr=1_5_9__205, rename the main .pck to \"yolkheroes.pck\" and add to the yolkheroes/gamedata/ directory. See https://portmaster.games/gptokeyb-documentation.html for details about Keyboard Text Entry.",
```
Comment (JeodC) [inline ports/yolkheroes/port.json:16]: From steam or itch.io, you can leave the urls out.

File: ports/yolkheroes/port.json [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+    "inst_md": null,
+    "genres": [
+      "simulation"
+    ],
+    "image": null,
+    "rtr": false,
+    "exp": false,
+    "runtime": "frt_3.5.2.squashfs",
+    "reqs": [
+      "hires"
```
Comment (JeodC) [inline ports/yolkheroes/port.json:26]: ?? It does?

**Review by JeodC — CHANGES_REQUESTED** [review]

File: ports/yolkheroes/Yolk Heroes.sh (hunk 3 above)
Comment (EzDzzIt) [reply ports/yolkheroes/Yolk Heroes.sh:37]: text input should be used, only to enter file name though

File: ports/yolkheroes/yolkheroes/yolkheroes.gptk (hunk 7 above)
Comment (EzDzzIt) [reply ports/yolkheroes/yolkheroes/yolkheroes.gptk:8]: I must have swapped this for testing, will fix

File: ports/yolkheroes/README.md [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+|--|--| 
+|A|'Confirm'|
+|B|'Back'|
+|X|'Confirm'|
+|Y|'Back'|
+|Start|'Menu'|
+|Select|'Space'|
+
+Please see https://github.com/PortsMaster/gptokeyb for a guide on text input.
```
Comment (JeodC) [inline ports/yolkheroes/README.md:20]: It's simple enough to just add a few lines to tell the user how to use text input. They're far less likely to get past the name screen if they have to go to another page to find out what to do.

File: ports/yolkheroes/Yolk Heroes.sh [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+elif [ -d "/opt/tools/PortMaster/" ]; then
+  controlfolder="/opt/tools/PortMaster"
+elif [ -d "$XDG_DATA_HOME/PortMaster/" ]; then
+  controlfolder="$XDG_DATA_HOME/PortMaster"
+else
+  controlfolder="/roms/ports/PortMaster"
+fi
+
+source $controlfolder/control.txt
+source $controlfolder/device_info.txt
```
Comment (JeodC) [inline ports/yolkheroes/Yolk Heroes.sh:16]:
```suggestion
```

File: ports/yolkheroes/Yolk Heroes.sh [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+$ESUDO mkdir -p "$godot_dir"
+$ESUDO umount "$godot_file" || true
+$ESUDO mount "$godot_file" "$godot_dir"
+PATH="$godot_dir:$PATH"
+
+export FRT_NO_EXIT_SHORTCUTS=FRT_NO_EXIT_SHORTCUTS
+
+$ESUDO chmod 666 /dev/uinput
+$GPTOKEYB "$runtime" -c "./yolkheroes.gptk" &
+SDL_GAMECONTROLLERCONFIG="$sdl_controllerconfig" "$runtime" $GODOT_OPTS --main-pack "gamedata/yolkheroes.pck"
```
Comment (JeodC) [inline ports/yolkheroes/Yolk Heroes.sh:63]:
```suggestion
pm_platform_helper "$runtime"
SDL_GAMECONTROLLERCONFIG="$sdl_controllerconfig" "$runtime" $GODOT_OPTS --main-pack "gamedata/yolkheroes.pck"
```

File: ports/yolkheroes/Yolk Heroes.sh [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+
+$ESUDO umount "$godot_dir"
```
Comment (JeodC) [inline ports/yolkheroes/Yolk Heroes.sh:65]:
```suggestion
pm_finish
```

`pm_finish` now takes care of all this cleanup so you really don't even need this block anymore.

**Review by JeodC — CHANGES_REQUESTED** [review]

File: ports/yolkheroes/LICENSE.FRT.txt [hunk 16]
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
```
Comment (JeodC) [inline ports/yolkheroes/LICENSE.FRT.txt:1]: This file needs to go inside the next yolkheroes folder, it's why the pr is failing on build

**Review by JeodC — CHANGES_REQUESTED** [review]

File: ports/yolkheroes/Yolk Heroes.sh [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+godot_dir="$HOME/godot"
+godot_file="$controlfolder/libs/${runtime}.squashfs"
+$ESUDO mkdir -p "$godot_dir"
+$ESUDO umount "$godot_file" || true
+$ESUDO mount "$godot_file" "$godot_dir"
+PATH="$godot_dir:$PATH"
+
+export FRT_NO_EXIT_SHORTCUTS=FRT_NO_EXIT_SHORTCUTS
+
+$ESUDO chmod 666 /dev/uinput
```
Comment (cdeletre) [inline ports/yolkheroes/Yolk Heroes.sh:60]: Not needed anymore
```suggestion
```

File: ports/yolkheroes/Yolk Heroes.sh [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+$GPTOKEYB "$runtime" -c "./yolkheroes.gptk" &
+pm_platform_helper "$runtime"
+SDL_GAMECONTROLLERCONFIG="$sdl_controllerconfig" "$runtime" $GODOT_OPTS --main-pack "gamedata/yolkheroes.pck"
+
```
Comment (cdeletre) [inline ports/yolkheroes/Yolk Heroes.sh:64]: Unmount is missing at the end
```suggestion
$ESUDO umount "$godot_dir"
```

**Review by cdeletre — CHANGES_REQUESTED** [review]

File: ports/yolkheroes/Yolk Heroes.sh (hunk 18 above)
Comment (kloptops) [reply ports/yolkheroes/Yolk Heroes.sh:64]: If you're doing this you might need to do:
```sh
if [[ "$PM_CAN_MOUNT" != "N" ]]; then
  $ESUDO umount "$godot_dir"
fi
```

File: ports/yolkheroes/LICENSE.FRT.txt (hunk 16 above)
Comment (JeodC) [inline ports/yolkheroes/LICENSE.FRT.txt:1]: This file needs to go inside the next yolkheroes folder, it's why the pr is failing on build

**Review by JeodC — CHANGES_REQUESTED** [review]

File: ports/yolkheroes/Yolk Heroes.sh [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+pm_finish
+$ESUDO kill -9 $(pidof gptokeyb)
+$ESUDO systemctl restart oga_events &
+printf "\033c" > /dev/tty0
```
[the remark this answers is not in the document]
Comment (EzDzzIt) [reply ports/yolkheroes/Yolk Heroes.sh:68]: that's nice!

**Review by cdeletre — APPROVED** [review]

**Review by JeodC — APPROVED** [review]

**Pull request merged**